File: 프로그래머스/3/42861. 섬 연결하기/섬 연결하기.js
function solution(n, costs) {
    let answer = 0;
    let edgy = 0;
    let parent = Array.from({length: n}, (_, i)=>i);
    
    costs.sort((a,b)=>a[2]-b[2]);
    
    // Kruskal
    for(const [a, b, cost] of costs){
        if(union(a, b)){
            edgy++;
            answer += cost;
        }
        
        if(edgy===n-1)  return answer
    }
    
    function union(a, b){
        const pA = find(a);
        const pB = find(b);

        if(pA===pB) return false;
        parent[pB] = pA;
        return true;
    }

    function find(x){
        if(parent[x]===x)   return x;
        parent[x] = find(parent[x]);
        return parent[x];
    }
    
    
    return answer;
}


// Kruskal Algorithm 를 사용해야함 -> 가중치 정렬해서 싸이클 안생기는 선에서 하나씩 선택
// http://chanhuiseok.github.io/posts/algo-33/
// https://lazyhysong.tistory.com/entry/JS-Kruskal-%EC%84%AC-%EC%97%B0%EA%B2%B0%ED%95%98%EA%B8%B0

File: 프로그래머스/3/42861. 섬 연결하기/섬 연결하기.py
def solution(n, costs):
    parent = [i for i in range(n)]
    
    def find(x):
        if parent[x] != x:
            parent[x] = find(parent[x])  # 경로 압축
        return parent[x]
    
    def union(a, b):
        root_a = find(a)
        root_b = find(b)
        
        if root_a != root_b:
            parent[root_b] = root_a
            return True
        return False
    
    costs.sort(key=lambda x: x[2])  # 비용 기준 정렬
    
    answer = 0
    count = 0
    
    for a, b, cost in costs:
        if union(a, b):   # 서로 다른 집합이면 연결
            answer += cost
            count += 1
            
            if count == n - 1:
                break
    
    return answer